README.txt
infrae.maildrophost
===================

``infrae.maildrophost`` is used to configure a maildrophost server and
`MaildropHost`_ product using the same configuration, and create a
management script for the maildrophost server.

In addition to those tasks, it used to download and install
`MaildropHost`_ when it was not distributed as an egg. If you are
looking for those features, please have a look at the version 1.x of
this recipe.

Example in buildout::

  [buildout]
  parts =
      maildrophost
      instance

  [maildrophost]
  recipe = infrae.maildrophost
  smtp_host = localhost
  smtp_port = 25

  [instance]
  ...
  eggs +=
     Products.MaildropHost
  zope-conf-additional +=
  <product-config maildrophost>
    config-path-application ${maildrophost:configuration}
  </product-config>


This will create the configuration file ``maildrophost.cfg`` for the
daemon, and put a start/stop script in the ``bin`` directory of the
buildout tree.

Spool and PID files are put by default in the ``var/maildrop``
directory, so data is preserved when update (if there is any data).

Settings
--------

You can customize some of settings of `MaildropHost`_:

``mail_dir``
  Directory to use as *home directory* for the deamon. By default it's
  ``${buildout:directory}/var/maildrop`` It will be created if it
  doesn't exists.

``spool_dir``
  Directory to use as a spool. By default it will be
  ``${mail_dir}/spool``. The directory will be created if it doesn't
  exist already.

``pid_file``
  PID file to use for the daemon. By default it will be
  ``${mail_dir}/maildrop.pid``.

``smtp_host``
  SMTP server to use. Default to localhost.

``smtp_port``
  Port to use. Default to default SMTP port, 25.

``login``
  If the SMTP server require authentication, login to use.

``password``
  If the SMTP server require authentication, password to use.

``tls``
  If 1, `MaildropHost`_ will be speaking to a TLS enabled SMTP server.

``batch``
  Set ``MAILDROP_BATCH``.

``add_messageid``
  Add a message id to the sent mail.

``pool_interval``
  Must be an integer which define the interval in seconds between two
  check for new mail in the spool directory. Default is 120 seconds.

``wait_interval``
  Must be an interger or float which say how much time the daemon
  should wait between sending two mails to the mail server.


Latest version
--------------

The latest version is available in a `Subversion repository
<https://svn.infrae.com/buildout/infrae.maildrophost/trunk>`_.

.. _MaildropHost: http://www.dataflake.org/software/maildrophost


docs/HISTORY.txt

History
=======

2.3 (unreleased)
----------------

* ...

2.2 (2013-10-07)
----------------

* Add ``fg`` command line option to start the script on the
  foreground.  For best results, combine this with enabling the
  ``supervised_daemon`` option. Patch provided by Maurits van Rees
  from Zest Software.


2.1 (2013-08-19)
----------------

* Update recipe to include the deamon script itself and update it to
  work with Python 2.6 and 2.7. This script used to require Python
  2.4.

* Update the start/stop script to be more reliable with the help of
  `psutil`_.

.. _psutil: https://pypi.python.org/pypi/psutil


2.0 (2012-08-23)
----------------

* Update recipe to work with the egg distribution of MaildropHost,
  still creating the configuration and start/stop script for the
  daemon in buildout.

1.2 (2009-02-13)
----------------

* Fix ``download-cache`` vs ``download-directory``. Don't assume the
  user already set it.

* Add options ``wait_interval``, ``add_messageid``, ``batch``,
  ``tls``, ``pid_file``, ``spool_dir``, ``mail_dir``.

* Remove the relative to buildout directory option ``mail-dir`` which
  was almost useless.

* Improve the control script.

1.1 (2009-01-23)
----------------

* Added support for MaildropHost >= 1.22 (added a version option, so
  the recipe can guess the config file name)

1.0.1
-----

* Fix the config file for Windows paths.

1.0
---

* Define the ``location`` in options, so you could refer to the
  product folder with ``${maildrophost:location}`` for exemple.

* Add the ``target`` option to use a specific destination folder for
  the product.

* Add maildrophost options ``login``, ``password``, and
  ``poll_interval``.

0.1
---

* First release, download and install MaildropHost product, make a
  ``config.py`` and create simple script for start and stop purpose.


infrae/maildrophost/ctl.py
# -*- coding: utf-8 -*-
# Copyright (c) 2007-2013 Infrae. All rights reserved.
# See also LICENSE.txt
# $Id$

import atexit
import os
import signal
import subprocess
import sys
import psutil

from maildrop import write_pid, exit_function, handle_sigterm
from stringparse import parse_assignments
from stringparse import ParserSyntaxError

SCRIPT = os.path.join(os.path.dirname(__file__), 'maildrop.py')


def is_process_running(pid):
    """Return if a process with a given PID is actually running
    """
    try:
        p = psutil.Process(pid)
        return p.is_running()
    except (psutil.NoSuchProcess, ValueError):
        return False


def maildrop_pid(pidfile):
    """Return the PID in pidfile if any
    """
    try:
        with open(pidfile) as pf:
            pid = int(pf.read())
            return pid
    ## file errors
    except IOError:
        print>>sys.stdout, 'No pidfile found.'
    except ValueError:
        print>>sys.stdout, 'Invalid pidfile, file will be deleted.'
        ## remove pidfile
        try:
            os.unlink(pidfile)
        except OSError:
            ## pidfile could have already been deleted, be silent.
            pass
    return -1


def maildrop_start(configuration, pidfile):
    """Start maildrophost.
    """
    if not os.path.isfile(SCRIPT):
        print>>sys.stderr, 'Could not find MaildropHost server script.'
        sys.exit(1)

    if not os.path.isfile(configuration):
        print>>sys.stderr, 'Could not find MaildropHost configuration.'
        sys.exit(1)

    ## if there's no running process then start one
    pid = maildrop_pid(pidfile)
    if not is_process_running(pid):
        psutil.Popen([sys.executable, SCRIPT, configuration])
        print>>sys.stdout, 'MaildropHost STARTED.'
    else:
        print>>sys.stderr, 'MaildropHost is already running with PID %s.' % pid
        sys.exit(1)


def maildrop_fg(configuration, pidfile):
    """Start maildrophost in the foreground.

    If you use this *and* you set DEBUG or SUPERVISED_DAEMON to a true
    value, it should be a fine configuration to use with supervisor.
    """
    if not os.path.isfile(SCRIPT):
        print>>sys.stderr, 'Could not find MaildropHost server script.'
        sys.exit(1)

    if not os.path.isfile(configuration):
        print>>sys.stderr, 'Could not find MaildropHost configuration.'
        sys.exit(1)

    ## if there's no running process then start one
    pid = maildrop_pid(pidfile)
    if not is_process_running(pid):
        print>>sys.stdout, 'Starting MaildropHost...'
        # We will not fork, so write the pid of the current script in
        # the pidfile.

        try:
            config = dict(parse_assignments(open(configuration).read()))
        except ParserSyntaxError:
            print>>sys.stderr, 'Cannot load config from "%s"' % configuration
            sys.exit(1)
        if not config['SUPERVISED_DAEMON']:
            print>>sys.stderr, (
                'When running in foreground mode SUPERVISED_DAEMON '
                'must be turned on in the configuration file at "%s"' %
                configuration)
            sys.exit(1)

        write_pid(pidfile, os.getpid())
        atexit.register(exit_function, pidfile)
        signal.signal(signal.SIGTERM, handle_sigterm)
        subprocess.call([sys.executable, SCRIPT, configuration])
    else:
        print>>sys.stderr, 'MaildropHost is already running with PID %s.' % pid
        sys.exit(1)


def maildrop_stop(configuration, pidfile):
    """Stop maildrophost.
    """
    pid = maildrop_pid(pidfile)
    try:
        p = psutil.Process(pid)
    except (psutil.NoSuchProcess, ValueError):
        print>>sys.stderr, 'MaildropHost is probably NOT running.'
        sys.exit(1)
    p.terminate()
    print>>sys.stdout, 'MaildropHost STOPPED.'
    try:
        p.wait(2)
    except psutil.TimeoutExpired:
        print>>sys.stdout, 'Termination timed out, process will be killed.'
        p.kill()
        print>>sys.stdout, 'MaildropHost KILLED.'
    ## remove pidfile
    try:
        os.unlink(pidfile)
    except OSError:
        ## pidfile could have already been deleted, be silent.
        pass


def maildrop_status(configuration, pidfile):
    """Look after the MaildropHost process status.
    """
    pid = maildrop_pid(pidfile)
    if is_process_running(pid):
        print>>sys.stdout, 'MaildropHost is running with PID %s' % pid
    else:
        print>>sys.stderr, 'MaildropHost is probably NOT running.'
        sys.exit(1)


def usage():
    print "usage: %s [start|fg|stop|restart|status]" % sys.argv[0]
    sys.exit(-255)


def main(options):
    if len(sys.argv) != 2:
        return usage()
    action = sys.argv[1]
    if action == 'start':
        return maildrop_start(**options)
    elif action == 'fg':
        return maildrop_fg(**options)
    elif action == 'stop':
        return maildrop_stop(**options)
    elif action == 'restart':
        maildrop_stop(**options)
        return maildrop_start(**options)
    elif action == 'status':
        return maildrop_status(**options)
    return usage()


infrae/maildrophost/maildrop.py
#!/usr/bin/env python
##############################################################################
#
# Copyright (c) 2004-2009 Jens Vagelpohl and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" maildrop daemon: delivers mail content produced by the MaildropHost

$Id: maildrop.py 1745 2009-03-23 11:02:22Z jens $
"""

import atexit
from email.Utils import make_msgid
import os
import rfc822
import signal
import smtplib
from stringparse import parse_assignments
from stringparse import ParserSyntaxError
import sys
import time

FATAL_ERROR_CODES = ('500', '501', '502', '503', '504', '550', '551', '553')


class MaildropError(Exception):
    """Maildrop Error.
    """
    pass


def usage():
    print 'Usage: maildrop.py /path/to/config'


def mainloop():
    while 1:
        # Are there any files in the spool directory?
        to_be_sent = []
        all_files = []

        for spool in MAILDROP_SPOOLS:
            all_files.extend([os.path.join(spool, x) for x in os.listdir(spool)])

        # Remove lock files
        clean_files = [x for x in all_files
                         if not x.endswith('.lck')]

        # Remove files that were locked by the previously removed lock files
        clean_files = [x for x in clean_files
                         if not '%s.lck' % x in all_files]

        # Remove directories
        to_be_sent = [x for x in clean_files if not os.path.isdir(x)]

        if len(to_be_sent) > 0:
            # Open the log file
            time_stamp = time.strftime('%Y/%m/%d %H:%M:%S')
            log_file = open(MAILDROP_LOG_FILE, 'a')
            msg = '\n### Started at %s...' % time_stamp
            log_file.write(msg)
            if DEBUG: print msg

            while len(to_be_sent) > 0:
                if (MAILDROP_BATCH == 0) or (MAILDROP_BATCH > len(to_be_sent)):
                    batch = len(to_be_sent)
                else:
                    batch = MAILDROP_BATCH

                # Send mail
                try:
                    smtp_server = smtplib.SMTP(SMTP_HOST, SMTP_PORT)
                    #smtp_server.set_debuglevel(1)
                    smtp_server.ehlo()
                except smtplib.SMTPConnectError:
                    # SMTP server did not respond. Log it and stop processing.
                    time_stamp = time.strftime('%Y/%m/%d %H:%M:%S')
                    err_msg = '!!!!! Connection error at %s' % time_stamp
                    finish_msg = '### Finished at %s' % time_stamp
                    log_file.write(err_msg)
                    if DEBUG: print err_msg
                    log_file.write(finish_msg)
                    if DEBUG: print finish_msg
                    break

                if MAILDROP_TLS > 0:
                    if (MAILDROP_TLS > 1 and
                         not smtp_server.has_extn('starttls')):
                        # Problem: TLS is required but the server does not
                        # offer it We stop processing here.
                        time_stamp = time.strftime('%Y/%m/%d %H:%M:%S')
                        err_msg = '!!!!! TLS unavailable at %s' % time_stamp
                        finish_msg = '### Finished at %s' % time_stamp
                        log_file.write(err_msg)
                        if DEBUG: print err_msg
                        log_file.write(finish_msg)
                        if DEBUG: print finish_msg
                        break

                    smtp_server.starttls()

                    # We have to say Hello again after starting TLS
                    smtp_server.ehlo()

                if MAILDROP_LOGIN != '' and MAILDROP_PASSWORD != '':
                    # Login is required to send mail
                    if not smtp_server.has_extn('auth'):
                        # The server does not offer authentication but we want it
                        # We stop processing here.
                        time_stamp = time.strftime('%Y/%m/%d %H:%M:%S')
                        err_msg = '!!!!! Authentication unavailable at %s' % time_stamp
                        finish_msg = '### Finished at %s' % time_stamp
                        log_file.write(err_msg)
                        if DEBUG: print err_msg
                        log_file.write(finish_msg)
                        if DEBUG: print finish_msg
                        break

                    try:
                        smtp_server.login(MAILDROP_LOGIN, MAILDROP_PASSWORD)
                    except smtplib.SMTPAuthenticationError:
                        # Authentication with the given credentials fails.
                        # We stop processing here.
                        time_stamp = time.strftime('%Y/%m/%d %H:%M:%S')
                        err_msg = '!!!!! Authentication failed at %s' % time_stamp
                        finish_msg = '### Finished at %s' % time_stamp
                        log_file.write(err_msg)
                        if DEBUG: print err_msg
                        log_file.write(finish_msg)
                        if DEBUG: print finish_msg
                        break

                for file_path in to_be_sent[0:batch]:
                    mail_dict = read_mail(file_path)
                    if not mail_dict:
                        continue

                    # Create mail and send it off
                    h_from = mail_dict.get('From')
                    h_to = mail_dict.get('To')
                    h_to_list = []
                    for item in rfc822.AddressList(h_to):
                        h_to_list.append(item[1])
                    h_body = mail_dict.get('body')

                    if ADD_MESSAGEID:
                        h_body = 'Message-Id: %s\n%s' % (make_msgid(), h_body)

                    try:
                        smtp_server.sendmail(h_from, h_to_list, h_body)
                        stat = 'OK'
                        os.remove(file_path)
                    except smtplib.SMTPRecipientsRefused, e:
                        stat = 'FATAL: ', str(e)
                        for (addr, error) in e.recipients.items():
                             if str(error[0]) in FATAL_ERROR_CODES:
                                 os.remove(file_path)
                                 break
                    except smtplib.SMTPException, e:
                        stat = 'BAD: ', str(e)

                    mail_msg = '\n%s\t %s' % (stat, h_to)
                    log_file.write(mail_msg)
                    if DEBUG: print mail_msg
                    log_file.flush()
                    if WAIT_INTERVAL:
                        time.sleep(WAIT_INTERVAL)

                to_be_sent = to_be_sent[batch:]

                try:
                    smtp_server.quit()
                except smtplib.SMTPServerDisconnected:
                    pass

            time_stamp = time.strftime('%Y/%m/%d %H:%M:%S')
            finish_msg = '\n### Finished at %s\n' % time_stamp
            log_file.write(finish_msg)
            if DEBUG: print finish_msg
            log_file.close()

        time.sleep(MAILDROP_INTERVAL)


def read_mail(file_path):
    """ Reads in a mail from a file, returns a dictionary with keys
    for headers and body
    """
    # Read in file
    file_handle = open(file_path, 'r')
    file_contents = file_handle.read()
    file_handle.close()

    # Is this a real mail turd?
    if not file_contents.startswith('##To:'):
        return

    # Parse and handle content (mail it out)
    mail_dict = {}
    file_lines = file_contents.split('\n')

    for i in range(len(file_lines)):
        if file_lines[i].startswith('##'):
            header_line = file_lines[i][2:]
            header_key, header_val = header_line.split(':', 1)
            mail_dict[header_key] = header_val
        else:
            mail_dict['body'] = '\n'.join(file_lines[i:])
            break

    return mail_dict


def write_pid(pidfile_path, pid):
    """ Write the daemon pid to our pid file """
    pid_file = open(pidfile_path, 'w')
    pid_file.write(str(pid))
    pid_file.close()


def exit_function(pidfile_path):
    # Remove the daemon pid file
    try:
        os.unlink(pidfile_path)
    except:
        pass


def handle_sigterm(signum, frame):
    sys.exit(0)


if __name__ == "__main__":
    try:
        if len(sys.argv) < 2:
            usage()
            sys.exit(1)

        c_dir, c_file = os.path.split(sys.argv[1])

        try:
            c_file = open(sys.argv[1])
        except IOError:
            raise MaildropError('Cannot find config file "%s"' % c_file)

        try:
            config = dict(parse_assignments(c_file.read()))
        except ParserSyntaxError:
            raise MaildropError('Cannot load config from "%s"' % c_file)

        MAILDROP_HOME = config['MAILDROP_HOME']
        MAILDROP_INTERVAL = config['MAILDROP_INTERVAL']
        MAILDROP_BATCH = config['MAILDROP_BATCH']
        MAILDROP_TLS = config['MAILDROP_TLS']
        MAILDROP_LOGIN = config['MAILDROP_LOGIN']
        MAILDROP_PASSWORD = config['MAILDROP_PASSWORD']
        SMTP_HOST = config['SMTP_HOST']
        SMTP_PORT = config['SMTP_PORT']
        DEBUG = config['DEBUG']
        WAIT_INTERVAL = config['WAIT_INTERVAL']
        SUPERVISED_DAEMON = config['SUPERVISED_DAEMON']

        # Windows does not support the detaching process below, so it is
        # always forced into debug mode.
        if sys.platform.startswith('win'):
            DEBUG = 1

        MAILDROP_SPOOL = config.get('MAILDROP_SPOOL')
        if MAILDROP_SPOOL:
            MAILDROP_SPOOLS = [x.strip() for x in MAILDROP_SPOOL.split(';')]
        else:
            MAILDROP_SPOOLS = [os.path.join(MAILDROP_HOME, 'spool')]

        MAILDROP_VAR = config.get('MAILDROP_VAR',
                                  os.path.join(MAILDROP_HOME, 'var'))
        MAILDROP_LOG_FILE = config.get('MAILDROP_LOG_FILE',
                                       os.path.join(MAILDROP_VAR, 'maildrop.log'))
        MAILDROP_PID_FILE = config.get('MAILDROP_PID_FILE',
                                       os.path.join(MAILDROP_VAR, 'maildrop.pid'))

        ADD_MESSAGEID = config.get('ADD_MESSAGEID', False)

        for spool in MAILDROP_SPOOLS:
            if not os.path.isdir(spool):
                os.makedirs(spool)

        if not os.path.isdir(MAILDROP_VAR):
            os.makedirs(MAILDROP_VAR)

        try:
            mail_server = smtplib.SMTP(SMTP_HOST, SMTP_PORT)
            mail_server.ehlo()

            if MAILDROP_TLS > 1:
                mail_server.starttls()
                mail_server.ehlo()

            if MAILDROP_LOGIN != '' and MAILDROP_PASSWORD != '':
                mail_server.login(MAILDROP_LOGIN, MAILDROP_PASSWORD)

            mail_server.quit()
        except:
            if DEBUG: import traceback; traceback.print_exc()
            msg = 'Invalid SMTP server "%s:%d"' % (SMTP_HOST, SMTP_PORT)
            raise MaildropError(msg)

    except SystemExit:
        sys.exit(0)
    except:
        usage()
        print
        print 'An error occured, aborting...'
        print "%s: %s" % (sys.exc_type, sys.exc_value)
        print
        sys.exit(1)

    if not (DEBUG or SUPERVISED_DAEMON):
        # Do the Unix double-fork magic; see Stevens's book "Advanced
        # Programming in the UNIX Environment" (Addison-Wesley) for details
        # In DEBUG mode we do not fork/detach from the terminal!
        try:
            pid = os.fork()
            if pid > 0:
                # Exit first parent
                sys.exit(0)
        except OSError, e:
            print >>sys.stderr, "fork #1 failed: %d (%s)" % (
                e.errno, e.strerror)
            sys.exit(1)

        # Decouple from parent environment
        os.chdir("/")
        os.setsid()
        os.umask(0)

        # Do second fork
        try:
            pid = os.fork()
            if pid > 0:
                # Exit from second parent; print eventual PID before exiting
                if DEBUG:
                    print "maildrop daemon PID %d" % pid
                write_pid(MAILDROP_PID_FILE, pid)
                sys.exit(0)
        except OSError, e:
            print >>sys.stderr, "fork #2 failed: %d (%s)" % (
                e.errno, e.strerror)
            sys.exit(1)
        atexit.register(exit_function,MAILDROP_PID_FILE)
        signal.signal(signal.SIGTERM,handle_sigterm)
    elif DEBUG:
        print '*****          Starting in DEBUG mode           *****'
        print '***** All log messages are shown on the console *****'
        sys.stdout.flush()

    # Start the daemon main loop
    mainloop()


infrae/maildrophost/__init__.py
# -*- coding: utf-8 -*-
# Copyright (c) 2007-2009 Infrae. All rights reserved.
# See also LICENSE.txt
# $Id$

import os, shutil, tempfile, urllib2, urlparse
import zc.buildout, zc.recipe.egg
import setuptools

URL = ('http://www.dataflake.org/software/maildrophost/'
       'maildrophost_%s/MaildropHost-%s.tgz')

class Recipe(object):

    def __init__(self, buildout, name, options):
        self.buildout = buildout
        self.name = name
        self.options = options
        self.location = options.get(
            'target',
            os.path.join(
                self.buildout['buildout']['parts-directory'],
                self.name))
        options['location'] = self.location
        self.product_location = os.path.join(self.location, 'MaildropHost')
        self.url = options.get('url', None)
        if self.url is None:
            version = options.get('version')
            self.url = URL % (version, version)

        self.dont_create_directories = options.get(
            'dont_create_directories',
            'no').lower() in set(['yes', 'on' , 'true'])
        self.egg = zc.recipe.egg.Egg(buildout, options['recipe'], options)
        self.mail_dir = self.options.get('mail_dir', None)
        if not self.mail_dir:
            self.mail_dir = os.path.join(
                self.buildout['buildout']['directory'], 'var', 'maildrop',)

    def install(self):
        """Install the maildrophost server
        """
        download_dir = self.buildout['buildout'].get(
            'download-cache',
            os.path.join(self.buildout['buildout']['directory'], 'downloads'))

        if not os.path.isdir(download_dir):
            os.mkdir(download_dir)

        _, _, urlpath, _, _, _ = urlparse.urlparse(self.url)
        tmp = tempfile.mkdtemp('buildout-'+self.name)
        try:
            fname = os.path.join(download_dir, urlpath.split('/')[-1])
            if not os.path.exists(fname):
                f = open(fname, 'wb')
                try:
                    f.write(urllib2.urlopen(self.url).read())
                except:
                    os.remove(fname)
                    raise zc.buildout.UserError(
                        "Failed to download URL %s: %s" % (self.url, str(e)))
                f.close()

            setuptools.archive_util.unpack_archive(fname, tmp)
            files = os.listdir(tmp)
            if not os.path.isdir(self.location):
                os.mkdir(self.location)
            shutil.move(os.path.join(tmp, files[0]), self.product_location)
        finally:
            shutil.rmtree(tmp)

        return self.update()

    def _build_config(self):
        """Create the config file for the maildrop server ; Create
        directory used by the maildrop server.
        """

        spool_dir = self.options.get(
            'spool_dir', None) or os.path.join(self.mail_dir, 'spool')
        pid_file = self.options.get(
            'pid_file', None) or os.path.join(self.mail_dir, 'maildrop.pid')
        if not self.dont_create_directories:
            if not os.path.exists(self.mail_dir):
                os.makedirs(self.mail_dir)
            if not os.path.exists(spool_dir):
                os.makedirs(spool_dir)

        config_option = dict(
            smtp_host=self.options.get('smtp_host', 'localhost'),
            smtp_port=self.options.get('smtp_port', '25'),
            maildrop_dir=self.mail_dir,
            spool_dir=spool_dir,
            pid_file=pid_file,
            executable=self.buildout['buildout']['executable'],
            login=self.options.get('login', ''),
            password=self.options.get('password', ''),
            poll_interval=self.options.get('poll_interval', '120'),
            wait_interval=self.options.get('wait_interval', '0.0'),
            add_messageid=self.options.get('add_messageid', '0'),
            supervised_daemon=self.options.get('supervised_daemon', '0'),
            batch=self.options.get('batch', '0'),
            tls=self.options.get('tls', '0'),)

        version = self.options.get('version', '0.00')
        major, minor = version.split('.')
        if int(major) < 2 and int(minor) < 22:
            config_filename = os.path.join(self.product_location, 'config.py')
        else:
            config_filename = os.path.join(self.product_location, 'config')
        config = open(config_filename, 'wb')
        config.write(maildrop_config_template % config_option)

    def _build_script(self):
        """Create the startup script in the bin directory.
        """
        requirements, ws = self.egg.working_set(['infrae.maildrophost'])
        pidfile=self.options.get(
            'pid_file', None) or os.path.join(self.mail_dir, 'maildrop.pid')
        config = dict(base=self.product_location,
                      pidfile=pidfile)

        zc.buildout.easy_install.scripts(
            [(self.name, 'infrae.maildrophost.ctl', 'main')],
            ws, self.options['executable'], self.options['bin-directory'],
            arguments=('%r, sys.argv[1:]' % config))

    def update(self):
        """Update the maildrophost server
        """
        try:
            self._build_config()
            self._build_script()
        except:
            shutil.rmtree(self.location)
            raise

        return self.location

maildrop_config_template="""
PYTHON=r"%(executable)s"
MAILDROP_HOME=r"%(maildrop_dir)s"
MAILDROP_VAR=r"%(maildrop_dir)s"
MAILDROP_SPOOL=r"%(spool_dir)s"
MAILDROP_PID_FILE=r"%(pid_file)s"

SMTP_HOST="%(smtp_host)s"
SMTP_PORT=%(smtp_port)s

MAILDROP_INTERVAL=%(poll_interval)s
DEBUG=0
DEBUG_RECEIVER=""

MAILDROP_BATCH=%(batch)s
MAILDROP_TLS=%(tls)s

MAILDROP_LOGIN="%(login)s"
MAILDROP_PASSWORD="%(password)s"

WAIT_INTERVAL=%(wait_interval)s
ADD_MESSAGEID=%(add_messageid)s

SUPERVISED_DAEMON=%(supervised_daemon)s
"""
